src/datasets/packaged_modules/imagefolder/imagefolder.py

```diff
@@ -119,44 +170,167 @@ def _split_files_and_archives(self, data_files):
+                                metadata_dir = os.path.dirname(metadata_file)
+                                metadata_dict = {
+                                    file_name: image_metadata
+                                    for file_name, image_metadata in zip(
+                                        pa_file_name_array.to_pylist(), pa_table_to_pylist(pa_metadata_table)
+                                    )
+                                }
+                        if metadata_dir is not None:
+                            file_relpath = os.path.relpath(file, metadata_dir)
+                            file_relpath = file_relpath.replace("\\", "/")
+                            image_metadata = metadata_dict.get(file_relpath, image_empty_metadata)
+                        else:
+                            image_metadata = image_empty_metadata
+                        if self.config.drop_labels:
+                            yield file_idx, {
+                                "image": downloaded_file_or_dir,
+                                **image_metadata,
+                            }
+                        else:
+                            yield file_idx, {
+                                "image": downloaded_file_or_dir,
+                                "label": os.path.basename(os.path.dirname(file)),
+                                **image_metadata,
+                            }
                         file_idx += 1
                 else:
                     for downloaded_dir_file in downloaded_file_or_dir:
                         _, downloaded_dir_file_ext = os.path.splitext(downloaded_dir_file)
                         if downloaded_dir_file_ext.lower() in self.IMAGE_EXTENSIONS:
-                            yield file_idx, {
-                                "image": downloaded_dir_file,
-                                "label": os.path.basename(os.path.dirname(downloaded_dir_file)),
-                            }
+                            current_dir = os.path.dirname(downloaded_dir_file)
+                            if last_checked_dir is None or last_checked_dir != current_dir:
+                                last_checked_dir = current_dir
+                                metadata_dir = None
+                                metadata_dict = None
+                                metadata_file_candidates = [
+                                    (
+                                        os.path.relpath(
+                                            downloaded_dir_file, os.path.dirname(downloaded_metadata_file)
+                                        ),
+                                        metadata_file,
+                                        downloaded_metadata_file,
+                                    )
+                                    for metadata_file, downloaded_metadata_file in metadata_files
+                                    if metadata_file is None
```

Comment:
If I understand correctly, checking that it is None here means that it's a single file, not coming from an archive.
I feel like it's a bit hard to understand that. Maybe separating single files from files from archives more explicitly would make the code more readable.